Modules/_ctypes/_ctypes.c

```diff
@@ -956,32 +956,51 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
         return NULL;
     }
 
+#undef USE_DLERROR
 #ifdef MS_WIN32
     Py_BEGIN_ALLOW_THREADS
     address = (void *)GetProcAddress(handle, name);
     Py_END_ALLOW_THREADS
-    if (!address) {
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-        return NULL;
-    }
 #else
+    #ifdef __CYGWIN__
+        // dlerror() isn't very helpful on cygwin
+    #else
+        #define USE_DLERROR
+        /* dlerror() always returns the latest error.
+         *
+         * Clear the previous value before calling dlsym(),
+         * to ensure we can tell if our call resulted in an error.
+         */
+        (void)dlerror();
+    #endif
     address = (void *)dlsym(handle, name);
-    if (!address) {
-#ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-#else
-        PyErr_SetString(PyExc_ValueError, dlerror());
 #endif
-        return NULL;
+
+    if (address) {
+        ctypes_state *st = get_module_state_by_def(Py_TYPE(type));
+        return PyCData_AtAddress(st, type, address);
+    }
+
+#ifdef USE_DLERROR
+    // This assumes the error message is UTF-8 (or ASCII).
+    // Investigate if this can cause problems.
+    const char *dlerr = dlerror();
+    if (dlerr) {
+        PyObject *message = PyUnicode_DecodeLocale(dlerr, "strict");
```

Comment:
I think we want to use `surrogateescape` instead of raising a `UnicodeError`:
```suggestion
        PyObject *message = PyUnicode_DecodeLocale(dlerr, "surrogateescape");
```

Reply:
Can you explain why?

Reply:
`strict` will raise a `UnicodeError` if it doesn't like something when decoding, I don't think that's a good idea when it comes to decoding the locale. For precedent, we also typically use `surrogateescape` when using `DecodeLocale` in most of the core, as far as I can see.